Gems/ROS2/Code/Source/Frame/ROS2FrameSystemComponent.cpp

```diff
@@ -0,0 +1,440 @@
+    {
+        return m_frameEntities.size();
+    }
+
+    ROS2FrameSystemComponent::ROS2FrameSystemComponent()
+    {
+        if (ROS2FrameSystemInterface::Get() == nullptr)
+        {
+            ROS2FrameSystemInterface::Register(this);
+        }
+    }
+
+    ROS2FrameSystemComponent::~ROS2FrameSystemComponent()
+    {
+        if (ROS2FrameSystemInterface::Get() == this)
+        {
+            ROS2FrameSystemInterface::Unregister(this);
+        }
+    }
+
+    void ROS2FrameSystemComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<ROS2FrameSystemComponent, AZ::Component>()->Version(0);
+        }
+    }
+
+    void ROS2FrameSystemComponent::Activate()
+    {
+    }
+
+    void ROS2FrameSystemComponent::Deactivate()
+    {
+    }
+
+    AZ::TransformInterface* ROS2FrameSystemComponent::GetEntityTransformInterface(const AZ::Entity* entity)
+    {
+        if (!entity)
+        {
+            AZ_Error("GetEntityTransformInterface", false, "Invalid entity!");
+            return nullptr;
+        }
+
+        auto* interface = Utils::GetGameOrEditorComponent<AzToolsFramework::Components::TransformComponent>(entity);
+
+        return interface;
+    }
+
+    bool ROS2FrameSystemComponent::IsTopLevel(const AZ::EntityId& frameEntityId) const
+    {
+        if (m_frameParent.contains(frameEntityId))
+        {
+            const AZ::EntityId& parent = m_frameParent.find(frameEntityId)->second;
+            return m_frameParent.find(parent)->second == parent;
+        }
+
+        return false;
+    }
+
+    AZ::EntityId ROS2FrameSystemComponent::GetParentEntityId(const AZ::EntityId& frameEntityId) const
+    {
+        if (m_frameParent.contains(frameEntityId))
+        {
+            return m_frameParent.find(frameEntityId)->second;
+        }
+        return AZ::EntityId();
+    }
+
+    AZStd::vector<AZ::EntityId> ROS2FrameSystemComponent::FindFrameParentPath(AZ::EntityId frameEntityId)
+    {
+        AZStd::vector<AZ::EntityId> path;
+        path.push_back(frameEntityId);
+
+        AZ::Entity* currentEntity = nullptr;
+        AZ::ComponentApplicationBus::BroadcastResult(currentEntity, &AZ::ComponentApplicationRequests::FindEntity, frameEntityId);
+        auto* currentTransform = GetEntityTransformInterface(currentEntity);
+        if (!currentTransform)
+        {
+            return path;
+        }
+        while (currentEntity->GetId().IsValid())
+        {
+            AZ::EntityId nextEntityId;
+            if (!currentTransform)
+            {
+                return path;
+            }
+            nextEntityId = currentTransform->GetParentId();
+            AZ::Entity* parent = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(parent, &AZ::ComponentApplicationRequests::FindEntity, nextEntityId);
+            if (parent == nullptr || !nextEntityId.IsValid())
```

Comment:
Is there a reason to not check nextEntityId before calling ComponentApplicationBus, that is probably quite expensive?